helix-term/src/commands/dap.rs

```diff
@@ -0,0 +1,445 @@
+    }
+}
+
+pub fn dap_next(cx: &mut Context) {
+    if let Some(debugger) = &mut cx.editor.debugger {
+        if debugger.is_running {
+            cx.editor
+                .set_status("Debuggee is already running".to_owned());
+            return;
+        }
+
+        let request = debugger.next(debugger.thread_id.unwrap());
+        if let Err(e) = block_on(request) {
+            cx.editor.set_error(format!("Failed to step: {:?}", e));
+        }
+    }
+}
+
+pub fn dap_variables(cx: &mut Context) {
+    if let Some(debugger) = &mut cx.editor.debugger {
+        if debugger.is_running {
+            cx.editor
+                .set_status("Cannot access variables while target is running".to_owned());
+            return;
+        }
+        let (frame, thread_id) = match (debugger.active_frame, debugger.thread_id) {
+            (Some(frame), Some(thread_id)) => (frame, thread_id),
+            _ => {
+                cx.editor
+                    .set_status("Cannot find current stack frame to access variables".to_owned());
+                return;
+            }
+        };
+
+        let frame_id = debugger.stack_frames[&thread_id][frame].id;
+        let scopes = match block_on(debugger.scopes(frame_id)) {
+            Ok(s) => s,
+            Err(e) => {
+                cx.editor
+                    .set_error(format!("Failed to get scopes: {:?}", e));
+                return;
+            }
+        };
+        let mut variables = Vec::new();
+
+        for scope in scopes.iter() {
+            let response = block_on(debugger.variables(scope.variables_reference));
+
+            if let Ok(vars) = response {
+                for var in vars {
+                    let prefix = match var.data_type {
+                        Some(data_type) => format!("{} ", data_type),
+                        None => "".to_owned(),
+                    };
+                    variables.push(format!("{}{} = {}\n", prefix, var.name, var.value));
+                }
+            }
+        }
+
+        if !variables.is_empty() {
+            cx.editor.variables = Some(variables);
+            cx.editor.variables_page = 0;
+        }
+    }
+}
+
+pub fn dap_terminate(cx: &mut Context) {
+    if let Some(debugger) = &mut cx.editor.debugger {
+        let request = debugger.disconnect();
+        if let Err(e) = block_on(request) {
+            cx.editor
+                .set_error(format!("Failed to disconnect: {:?}", e));
+            return;
+        }
+        cx.editor.debugger = None;
+    }
+}
+
+pub fn dap_switch_thread(cx: &mut Context) {
+    if let Some(debugger) = &mut cx.editor.debugger {
+        let request = debugger.threads();
+        let threads = match block_on(request) {
+            Ok(threads) => threads,
+            Err(e) => {
+                cx.editor
+                    .set_error(format!("Failed to retrieve threads: {:?}", e));
+                return;
+            }
+        };
+
+        let picker = Picker::new(
```

Comment:
TODO: Previews here if source code is available for that pointer

Reply:
Good point, I didn't consider that

Reply:
I'll do it!

Reply:
Changed my mind here, we don't want to use a picker on threads since they might be running, let's do this on stack frames instead 👍🏻